### ISSUE TITLE
Native flag ignored in 0.59.0

### ISSUE THREAD

**Issue by bentitmus** [mesonbuild/meson#9037]
**Describe the bug**
When updating to the latest version of `meson` it appears that the `native` flag is ignored in `add_project_arguments` (and possibly other places although I haven't checked).

**To Reproduce**
`meson.build`:
```
project('demo', 'c')
add_project_arguments('-DA', native: true, language: 'c')
add_project_arguments('-DB', native: false, language: 'c')
executable('demo', 'demo.c')
```

`demo.c`:
```
int main(void) {
  return 0;
}
```

Run: 
* `meson setup build`
* `cd build`
* `ninja -v`

This gives something like:
```
[1/2] ccache cc -Idemo.p -I. -I.. -fcolor-diagnostics -Wall -Winvalid-pch -O0 -g -DA -DB -MD -MQ demo.p/demo.c.o -MF demo.p/demo.c.o.d -o demo.p/demo.c.o -c ../demo.c
[2/2] cc  -o demo demo.p/demo.c.o -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,error
```

Which includes `-DA` and `-DB` although this is a native build.

**Expected behavior**
I expect only `-DA` to be included in the arguments as `-DB` should only be included on cross-compiles (as per the reference manual).

**system parameters**
* This is running as a native build on Linux and macOS
* Meson version is `0.59.0`
* Ninja version is `1.10.2` (although this isn't relevant because it's an error in the `build.ninja` file)
* Python is `3.8.2`

**Comment by bentitmus** [comment]
Works as expected in `0.58.2`.

**Comment by dcbaker** [comment]
This seems correct as is it me, if you're doing a "native" build then the machine is both the host and the build machine, (or it's both native and not native at the same time). But if this isn't the previous behavior... Sigh.

**Comment by bentitmus** [comment]
I'm actually slightly confused about this myself (which shows that perhaps the `native` flag isn't entirely obvious). From the reference manual:

> native (since 0.48.0): a boolean specifying whether the arguments should be applied to the native or cross compilation. If true the arguments will only be used for native compilations. If false the arguments will only be used in cross compilations. If omitted, the flags are added to native compilations if compiling natively and cross compilations (only) when cross compiling.

This suggests that `native: true` should indicate that it is present when `host` is the same as `build` ('native' compilations) and `native: false` indicates that it should be present only when `host` is not the same as `build` (cross compilations). In fact, this is what I see in my actual project builds.

When running the above simple example I came up with to reproduce the issue, I notice that it is instead only using the `native: false` argument when instead I would expect it to only use the `native: true` argument (in `0.58.2`).

**Comment by bentitmus** [comment]
Apologies for the confusion as it has been some time since I looked into cross compilation under meson. If I use the following:

```
project('demo', 'c')
add_project_arguments('-DA', native: true, language: 'c')
add_project_arguments('-DB', native: false, language: 'c')
executable('demo', 'demo.c', native: true)
```

Then under `0.58.2` it will use the native argument but not the non-native argument, which agrees with the documentation.

I believe the native flag is supposed to be used to indicate whether an executable or library is a tool to be used in the build process (`native: true`) or whether it should be executed on the final platform (which may be cross compiled: `native: false`). So project arguments with the `native` flag don't indicate whether `host` and `build` are the same, but whether they apply for build tools (`native: true`) which must be on the `build` platform, or whether they apply to the final products that will be on the host platform (`native: false`).

I slightly abused this in the project I wrote the `meson.build` for which is why I was confused.
